### PR TITLE
Fixed AttributeError when unpickling HTTPAdapterWithSocketOptions.

### PR DESCRIPTION
**Pull request by creste**
#### Pull Request Description

Fixes https://github.com/Azure/azure-kusto-python/issues/430

#### Future Release Comment
**Fixes:**
- Fixed Attribute error when unpickling HTTPAdapterWithSocketOptions

### CODE REVIEW

**Comment by creste** [comment]
@microsoft-github-policy-service agree [company="Anonym"]

**Comment by creste** [comment]
@microsoft-github-policy-service agree company="Anonym"

**Comment by AsafMah** [comment]
Great job!

What about the other attributes? don't they also need pickling?

**Comment by creste** [comment]
@AsafMah - Great question!

### pool_maxsize

`HTTPAdapterWithSocketOptions.pool_maxsize` does not need to be pickled because the base class, `HTTPAdapter`, uses and [pickles](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L158-L159) [`_pool_maxsize`](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L132) instead (note the underscore).  I'm _guessing_ `HTTPAdapterWithSocketOptions.pool_maxsize` exists to `pop()` `pool_maxsize` from `kwargs`, which forces `HTTPAdapter` to use the [`DEFAULT_POOLSIZE`](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L139).

Alternatively, I can remove this line of code:
```python
self.pool_maxsize = kwargs.pop("pool_maxsize", None)
```
But that means `HTTPAdapter` will no longer use the [`DEFAULT_POOLSIZE`](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L139).  I'm not sure if that will break something.

### max_retries
`HTTPAdapterWithSocketOptions.max_retries` is already [pickled](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L158-L159) by [`HTTPAdapter`](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L143-L146), so it doesn't need to be pickled by `HTTPAdapterWithSocketOptions.__getstate__()` again.

Similar to `pool_maxsize`, I can remove this line of code:
```python
self.max_retries = kwargs.pop("max_retries", None)
```
But that means `HTTPAdapter` will no longer use the default [`MAX_RETRIES`](https://github.com/psf/requests/blob/1e62a3ec18e19f85ddae03b4cbbdf0b4c62834c0/requests/adapters.py#L143-L144). I am not sure if that will break something.

**Comment by AsafMah** [comment]
Great point!
That probably explains #429, it should be `get` instead of `pop`.

**Comment by AsafMah** [comment]
Good point, yes, they could be just removed.

**Review by AsafMah — CHANGES_REQUESTED** [review]

File: azure-kusto-data/tests/test_http_adapter_with_socket_options.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License
+import pickle
+import unittest
+import socket
+
+from azure.kusto.data.client import HTTPAdapterWithSocketOptions
+
+class HTTPAdapterWithSocketOptionsTests(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (creste) [reply azure-kusto-data/tests/test_http_adapter_with_socket_options.py]: Updated to use pytest.

File: azure-kusto-data/tests/test_http_adapter_with_socket_options.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    def test_pickle(self):
+        adapter = HTTPAdapterWithSocketOptions(socket_options=[(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)])
+        pickled = pickle.dumps(adapter)
+        pickle.loads(pickled)
```
[the remark this answers is not in the document]
Comment (creste) [reply azure-kusto-data/tests/test_http_adapter_with_socket_options.py]: Added assert to verify `socket_options` are picked and unpickled correctly.

**Comment by AsafMah** [comment]
Awesome!

**Review by AsafMah — APPROVED** [review]

**Pull request merged**